### PR TITLE
🎉 GCS destination: use serialized buffer; compress csv & jsonl

### LINKED ISSUE #11425: Apply buffering changes to GCS Destination
## Tell us about the problem you're trying to solve
As part of https://github.com/airbytehq/airbyte/issues/10260, refactor destination-gcs to use serialized buffering strategy.

### PR DESCRIPTION
**Pull request by tuliren**
## What
- Cleaned up S3 destination config.
  - Add a builder.
  - Removed all redundant constructors.
- Refactored GCS destination config.
  - The GCS config now extends from an S3 config for convenience.
- Migrated the GCS destination to use the new serialized buffer strategy to be more scalable.
- This PR resolves https://github.com/airbytehq/airbyte/issues/11425.
- This PR relates to https://github.com/airbytehq/airbyte/issues/11203.

## Config class
- Each destination config class includes a `BlobStorageCredentialConfig<ConfigType>`
  - For S3, it is `S3CredentialConfig`, whose `ConfigType` is `S3CredentialType`.
  - For GCS, it is `GcsCredentialConfig`, whose `ConfigType` is `GcsCredentialType`.
- S3 has two credential config implementations
  - `S3AccessKeyCredentialConfig`
  - `S3InstanceProfileCredentialConfig`
- GCS has one credential config implementation
  - `GcsHmacKeyCredentialConfig`
- The GCS credential config can be converted to an S3 credential config.
  - This is because currently we always use the S3 client to communicate with GCS.
  - This may change in the future.

## Recommended reading order
1. `GcsDestination.java`
2. `GcsDestinatioConfig.java`
3. `S3DestinationConfig.java`

## 🚨 User Impact 🚨
1. The main improvement from this PR is that the GCS destination can handle a large number of streams now. Previously it will throw out-of-memory error whenever there are too many streams (e.g. 50).
2. 🚨 The impact for the users is that CSV and JSONL formats will be automatically compressed with GZIP. For example, if an output file was previously `abc.csv`, it will become `abc.csv.gz` after this PR.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<strong>Updating a connector</strong>

### Community member or Airbyter

- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [x] Code reviews completed
- [x] Documentation updated
    - [x] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [x] Create a non-forked branch based on this PR and test the below items on it
- [x] Build is successful
- [x] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [x] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

### CODE REVIEW

**Comment by tuliren** [comment]
@ChristopheDuong, this is the prerequisite for BigQuery changes. It updates many of the GCS classes to be compatible with S3 to directly reuse the S3 code. It also completes the migration for the GCS destination.

**Review by ChristopheDuong — APPROVED** [review]

File: airbyte-integrations/connectors/destination-gcs/src/main/java/io/airbyte/integrations/destination/gcs/GcsDestination.java [hunk 1]
```diff
@@ -27,6 +30,12 @@
   public static final String EXPECTED_ROLES = "storage.multipartUploads.abort, storage.multipartUploads.create, "
       + "storage.objects.create, storage.objects.delete, storage.objects.get, storage.objects.list";
 
+  private final NamingConventionTransformer nameTransformer;
+
+  public GcsDestination() {
+    this.nameTransformer = new S3NameTransformer();
```
[the remark this answers is not in the document]
Comment (tuliren) [reply airbyte-integrations/connectors/destination-gcs/src/main/java/io/airbyte/integrations/destination/gcs/GcsDestination.java:36]: Good point. Will do.

**Comment by tuliren** [comment]
Will publish the connector in a follow up PR.

**Comment by wallies** [comment]
Who decided this was a good idea for an ETL tool to output to gz. How are tools using this data afterwards meant to process it
```
The impact for the users is that CSV and JSONL formats will be automatically compressed with GZIP. For example, if an output file was previously abc.csv, it will become abc.csv.gz after this PR.
```

**Comment by tuliren** [comment]
@wallies, thank you for raising this question and creating [this issue](https://github.com/airbytehq/airbyte/issues/11872). We made the decision to compress CSV and JSONL formats based on the common use case of these blob storages. People usually use S3 and GCS just to archive their data, and the compression can reduce the storage cost.

I admit that this is not friendly for other use cases. I have created an issue [here](https://github.com/airbytehq/airbyte/issues/12001). Should have a new version that provides an option to not compress these formats by the end of this week or early next week.

**Comment by wallies** [comment]
> @wallies, thank you for raising this question and creating [this issue](https://github.com/airbytehq/airbyte/issues/11872). We made the decision to compress CSV and JSONL formats based on the common use case of these blob storages. People usually use S3 and GCS just to archive their data, and the compression can reduce the storage cost.
> 
> I admit that this is not friendly for other use cases. I have created an issue [here](https://github.com/airbytehq/airbyte/issues/12001). Should have a new version that provides an option to not compress these formats by the end of this week or early next week.

Much appreciated @tuliren. I also raised this https://github.com/airbytehq/airbyte/issues/11872 as we were seeing no file extension at all on any new files

**Comment by tuliren** [comment]
@wallies, the new version with an option to not compress CSV and JSONL files has been published. Please give the new version a try. Thanks~~

**Pull request merged**